nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
-            let results = Array::from_slice(agent, &slice);
+        // 3. Let S be ? ToString(O).
+        let s = to_string(agent, o)?;
 
-            return Ok(results.into_value());
-        }
+        // 4. If limit is undefined, lim is u32::MAX - 1.
+        let limit = args.get(0);
+        let lim = match limit {
+            Value::Undefined => u32::MAX,
+            _ => to_uint32(agent, limit)?,
+        };
 
-        let separator_str = to_string(agent, separator_value)?;
+        // 5. Let R be ? ToString(separator).
+        let r = to_string(agent, separator)?;
 
-        // 7. Let limit to be a number
-        let limit = to_number(agent, args.get(1))?;
+        // 6. If lim is zero, return an empty array
+        if lim == 0 {
+            let list: [Value; 0] = [];
+            return Ok(create_array_from_list(agent, &list).into_value());
+        }
 
-        // 8. If the limit is zero, returns an empty array
-        if limit.is_pos_zero(agent) || limit.is_neg_zero(agent) {
-            let a = array_create(agent, 0, 0, None)?;
-            return Ok(a.into_value());
+        // 7. If separator is undefined, return an array with the whole string
+        if let Value::Undefined = separator {
+            let list: [Value; 1] = [s.into_value()];
+            return Ok(create_array_from_list(agent, &list).into_value());
         }
 
-        // 9. Split the string
-        let subject = s.as_str(agent).to_owned();
-        let separator = separator_str.as_str(agent).to_owned();
-        let split = subject.split(&separator);
+        // 8. Let separatorLength be the length of R.
+        let separator_length = r.len(agent);
 
-        // 10. The limit must be positive integer or 0
-        let limit = if limit.is_sign_positive(agent) {
-            limit.into_i64(agent) as usize
-        } else {
-            0
-        };
+        // 9. If separatorLength = 0, the split by characters
+        if separator_length == 0 {
+            let subject = s.as_str(agent).to_owned();
+            let head = subject.split("");
+            let mut results: Vec<Value> = Vec::new();
 
-        // 11 Collect the results
-        let mut results: Vec<Value> = Vec::new();
-        for (i, part) in split.enumerate() {
-            if limit != 0 && limit <= i {
-                break;
+            for (i, part) in head.enumerate() {
+                println!("--->{}-{}", i, part);
+                if lim as usize == i {
+                    break;
+                }
+                results.push(Value::from_str(agent, part));
             }
-            results.push(Value::from_str(agent, part));
-        }
 
-        // 12. If separator is empty, Rust's split inserts an empty string in the beginning and end of the array
-        if separator.is_empty() {
+            // Note: Rust's split inserts an empty string in the beginning and end of the array. We remove them
             if results.len() > 1 {
                 results.remove(0);
```

Comment:
**suggestion:** It would be better to do this "filtering" out as part of the enumeration itself.

You can probably do something like: `let results: Vec<Value> = head.enumerate().skip(1).take_while(lim).map(|part| Value::from_str(agent, part)).collect();`. Now when creating this from slice, just drop the last item if the length of the Vec is longer than 0.

Reply:
done